src/Microsoft.DotNet.Interactive.CSharpProject/CSharpProjectKernel.cs

```diff
@@ -1,303 +1,318 @@
-
-            KernelEventEnvelope.RegisterEvent<DocumentOpened>();
-            KernelEventEnvelope.RegisterEvent<AssemblyProduced>();
-        }
-
-        public CSharpProjectKernel(string name)
-            : base(name)
-        {
-            RegisterCommandHandler<OpenProject>(HandleAsync);
-            RegisterCommandHandler<OpenDocument>(HandleAsync);
-            RegisterCommandHandler<CompileProject>(HandleAsync);
-            RegisterCommandHandler<RequestCompletions>(HandleAsync);
-            RegisterCommandHandler<RequestDiagnostics>(HandleAsync);
-            RegisterCommandHandler<RequestSignatureHelp>(HandleAsync);
-            RegisterCommandHandler<SubmitCode>(HandleAsync);
-        }
-
-        public async Task HandleAsync(OpenProject command, KernelInvocationContext context)
-        {
-            var package = await CreateConsoleWorkspacePackage();
-            _workspaceServer = new RoslynWorkspaceServer(package);
-
-            var extractor = new BufferFromRegionExtractor();
-            _workspace = extractor.Extract(command.Project.Files.Select(f => new Protocol.File(f.RelativeFilePath, f.Content)).ToArray());
-
-            context.Publish(new ProjectOpened(command, _workspace.Buffers.GroupBy(b => b.Id.FileName)
-                .OrderBy(g => g.Key).Select(g => new ProjectItem(g.Key, g.Select(r => r.Id.RegionName).Where(r => r != null).OrderBy(r => r).ToList())).ToList()));
-        }
-
-        public async Task HandleAsync(OpenDocument command, KernelInvocationContext context)
-        {
-            EnsureProjectOpened();
-
-            var file = _workspace.Files.SingleOrDefault(f => f.Name == command.RelativeFilePath);
-            if (file is null)
-            {
-                // check for a region-less buffer instead
-                var buffer = _workspace.Buffers.SingleOrDefault(b => b.Id.FileName == command.RelativeFilePath && b.Id.RegionName is null);
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Clockwise;
+using Microsoft.DotNet.Interactive.Commands;
+using Microsoft.DotNet.Interactive.CSharpProject.Commands;
+using Microsoft.DotNet.Interactive.CSharpProject.Events;
+using Microsoft.DotNet.Interactive.CSharpProject.MLS.Project;
+using Microsoft.DotNet.Interactive.CSharpProject.Packaging;
+using Microsoft.DotNet.Interactive.CSharpProject.Servers.Roslyn;
+using Microsoft.DotNet.Interactive.Events;
+using Microsoft.DotNet.Interactive.Server;
+
+namespace Microsoft.DotNet.Interactive.CSharpProject
+{
+    public class CSharpProjectKernel
+        : Kernel
+    {
+        private RoslynWorkspaceServer _workspaceServer;
+        private Protocol.Workspace _workspace;
+        private Protocol.Buffer _buffer;
+
+        static CSharpProjectKernel()

```

Comment:
I'd avoid doing this in a static constructor since these things can potentially be reset. The more explicit code is also more performant.

If there's a maintainability issue here let's fix the design bug that prevents `IKernelCommandHandler<T>` from working as expected.